opm/core/grid/RepairZCORN.hpp

```diff
@@ -0,0 +1,728 @@
+
+            for (const auto& globCell : this->active_.activeGlobal()) {
+                all_signs.push_back(this->getZCornSign(globCell));
+            }
+
+            const int ignore = 0;
+
+            // Elevation == ZCORN decreasing => all non-ignored cell signs
+            // negative.
+            return (first(all_signs, ignore) == -1)
+                && allEqual(all_signs, ignore, -1);
+        }
+
+        /// Retrieve sign of single cell's ZCORN change.
+        ///
+        /// \tparam CellIndex Integral type representing cell indices.
+        ///
+        /// \param[in] globCell Global ID of particular cell.
+        ///
+        /// \return Sign of \p globCell's ZCORN change.  Positive (+1) if
+        ///    ZCORN does not *DECREASE* along any of the cell's pillars,
+        ///    zero (0) if ZCORN increases along some of the pillars and
+        ///    decreases along others (or does not change at all), and
+        ///    negative (-1) if ZCORN does not *INCREASE* along any of the
+        ///    cell's pillars.
+        template <typename CellIndex>
+        int getZCornSign(const CellIndex globCell) const
+        {
+            auto sign = [](const double x) -> int
+            {
+                return (x > 0.0) - (x < 0.0);
+            };
+
+            const auto ijk = this->active_.getCellIJK(globCell);
+
+            auto sgn = std::vector<int>{};  sgn.reserve(4);
+
+            for (const auto& pt : this->zcorn_idx_.pillarPoints(ijk)) {
+                const auto dz =
+                    this->zcorn_[pt.bottom] - this->zcorn_[pt.top];
+
+                sgn.push_back(sign(dz));
+            }
+
+            const int ignore = 0;
+
+            if (! allEqual(sgn, ignore)) {
+                return 0;
+            }
+
+            return sgn.front();
+        }
+
+        /// Check whether or not all elements of a vector are equal, while
+        /// ignoring particular element value.
+        ///
+        /// \param[in] coll Sequence of elements.
+        ///
+        /// \param[in] ignore Special value that should be considered equal
+        ///    (i.e., ignored) for comparison purposes.  Typically zero.
+        ///
+        /// \return Whether or not all non-ignored elements of \p coll have
+        ///    the same value.
+        bool allEqual(const std::vector<int>& coll,
+                      const int               ignore) const
+        {
+            return this->allEqual(coll, ignore, ignore);
+        }
+
+        /// Check whether or not all elements of a vector are equal, while
+        /// ignoring particular element value.
+        ///
+        /// \param[in] coll Sequence of elements.
+        ///
+        /// \param[in] ignore Special value that should be considered equal
+        ///    (i.e., ignored) for comparison purposes.  Typically zero.
+        ///
+        /// \param[in] lookfor Search hint for which value to look for in
```

Comment:
> If calling `allEqual(v, ignore, x)` I am asking if all non-ignored elements of `v` are equal to `x`, not if they are all equal to each other.

Sure.  However, we're only using the three-parameter version of `allEqual()` if we already **know** what `first(v, ignore)` returns.  In the context of that use, we're omitting the first scan that would otherwise be needed to determine which value to look for.